ci-operator/step-registry/cucushift/installer/check/azure/usertags/cucushift-installer-check-azure-usertags-commands.sh

```diff
@@ -0,0 +1,253 @@
+if [[ -z "${RESOURCE_GROUP}" ]]; then
+    RESOURCE_GROUP="${INFRA_ID}-rg"
+fi
+
+check_result=0
+
+ocp_owned_key="kubernetes.io_cluster.${INFRA_ID}"
+ocp_owned_value="owned"
+user_tags_from_config=$(yq-go r ${INSTALL_CONFIG} 'platform.azure.userTags' -j | sed 's/false/"false"/g' | jq -c)
+echo "user_tags_from_config: ${user_tags_from_config}"
+#append OCP owned tag
+expected_user_tags=$(echo "${user_tags_from_config}" | jq -c -S ". += {\"${ocp_owned_key}\":\"${ocp_owned_value}\"}")
+echo "expected_user_tags: ${expected_user_tags}"
+
+#check tags on resource created in resource group
+resource_info_file=$(mktemp)
+echo "All resrouces created in cluster resource group ${RESOURCE_GROUP}:"
+az resource list -g ${RESOURCE_GROUP} --query '[].{type:type, name:name, tags:tags}' -ojson | tee ${resource_info_file}
+readarray -t res_info_array < <(jq -c '.[]' ${resource_info_file})
+echo -e "\n************ Checking resources tags created in resource group ${RESOURCE_GROUP} ************ "
+for item in "${res_info_array[@]}"; do
+    res_type=$(echo ${item} | jq -r -c '.type')
+    res_name=$(echo ${item} | jq -r -c '.name')
+    res_tags=$(echo ${item} | jq -r -c -S '.tags')
+
+    echo "(*) check tags on resource ${res_type} ${res_name}"
+    echo "tags on ${res_name}: ${res_tags}"
+
+    if [[ "${res_type}" == "Microsoft.Network/publicIPAddresses" ]] && [[ "${res_tags}" =~ k8s-azure-cluster-name ]]; then
+        echo "WARN: app public ip address is not tagged with openshift or user tags since it is not created by openshift core operator! Skip the check..."
+        continue
+    fi
+
+    if [[ "${res_type}" == "Microsoft.Compute/disks" ]] && [[ "${res_tags}" =~ kubernetes.io-created-for-pv-name ]]; then
+        echo "WARN: here is limitation, user-defined tags for PV volumnes cannot be updated! Skip the check..."
+        continue
+    fi
+
+    if [[ "${res_tags}" != "${expected_user_tags}" ]]; then
+        echo "ERROR: tags attached on resource ${res_name} don't match with expected tags list!"
+        check_result=1
+    else
+        echo "INFO: tags attached on resource ${res_name} match with expected tags list!"
+    fi
+done
+
+#check tags on resource group
+echo -e "\n************ Checking tags on resource group ${RESOURCE_GROUP} ************ "
+rg_tags=$(az group show --name ${RESOURCE_GROUP} --query tags -ojson | jq -c )
+rg_tags=$(echo "${rg_tags}" | jq -c -S 'del(.openshift_creationDate)')
+echo "tags on ${RESOURCE_GROUP}: ${rg_tags}"
+if [[ "${rg_tags}" != "${expected_user_tags}" ]]; then
+    echo "ERROR: tags attached on resource group ${RESOURCE_GROUP} don't match with expected tags list!"
+    check_result=1
+else
+    echo "INFO: tags attached on resource group ${RESOURCE_GROUP} match with expected tags list!"
+fi
+
+#check tags on public dns records 
+echo -e "\n************ Checking tags on dns records in public/private zone ************ "
+echo "(*) checking tags on app dns record in public zone"
```

Comment:
How about judging if the cluster is a public cluster, then decide if check the public resources, so that in some day, we can easily extend the testing to private cluster?

Reply:
good point, updated